### ISSUE TITLE
Regent - comparative performance of math.h calls vs C and/or accurately timing regent loops

### ISSUE THREAD

**Issue by LonelyCat124** [StanfordLegion/legion#848]
Hi,

I've been working on a large loop with a variety of if statements, trying to remove them to allow me to run with `__demand(__vectorize)` and I noticed a loss in performance when moving from some if statements to a math.h call : copysign.

To test this further, I tried creating a standalone code:
```
import "regent"


copysign = regentlib.copysign(double)

local c = regentlib.c
local format = require("std/format")

fspace point{
  a : double
}

task get_time() : double

  return c.legion_get_current_time_in_micros()
end

task main()

 var reg = region( ispace(int1d, 250000000), point)
 var reg2 = region( ispace(int1d, 250000000), point)

 fill(reg.a, 0.75)
 fill(reg2.a, -1.0)

 for point in reg do
    reg[point].a = copysign(reg[point].a, reg2[point].a)
 end

 fill(reg.a, 0.75)
 fill(reg2.a, -1.0)

  __fence(__execution, __block)
  var start_time = get_time()
  __fence(__execution, __block)

 __forbid(__vectorize)
 for point in reg do
    reg[point].a = copysign(reg[point].a, reg2[point].a)
 end
  __fence(__execution, __block)
  var finish_time = get_time()
  __fence(__execution, __block)
  var time_taken = finish_time - start_time
  format.println("non-vectorize copysign: {}", double(time_taken)/1000000.0)
  __fence(__execution, __block)
 fill(reg.a, 0.75)
 fill(reg2.a, -1.0)

  __fence(__execution, __block)
  start_time = get_time()
  __fence(__execution, __block)

 __demand(__vectorize)
 for point in reg do
    reg[point].a = copysign(reg[point].a, reg2[point].a)
 end
  __fence(__execution, __block)
  finish_time = get_time()
  __fence(__execution, __block)
  time_taken = finish_time - start_time
  format.println("vectorize copysign: {}", double(time_taken)/1000000.0)
  __fence(__execution, __block)

 fill(reg.a, 0.75)
 fill(reg2.a, -1.0)

  __fence(__execution, __block)
  start_time = get_time()
  __fence(__execution, __block)

  for point in reg do
    var z : double = 0.0
    if reg2[point].a >= 0.0 then
      z = 1.0
    else
      z = -1.0
    end
    reg[point].a = z * reg[point].a
  end

  __fence(__execution, __block)
  finish_time = get_time()
  __fence(__execution, __block)
  time_taken = finish_time - start_time
  format.println("if statement: {}", double(time_taken)/1000000.0)
  __fence(__execution, __block)

end
regentlib.start(main)
```

Without the warm-up loop, the performance is much slower for the first loop than the later ones. With this code, I see this performance (across many runs I see similar performance). Compiled with llvm 6, running on Skylake, ran with this command: `legion/language/regent.py test_sign.rg  -ll:csize 4096`:
```
non-vectorize copysign: 1.287462
vectorize copysign: 1.287276
if statement: 1.130615
```

I also created a C code to mimic this behaviour:
```
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <omp.h>

void fill(double *reg_a, double *reg_b){

  for( int i = 0; i < 25000000; i++){
    reg_a[i] = 0.75;
    reg_b[i] = -1.0;
  }

}

int main(){


  double *reg_a, *reg_b;

  reg_a = malloc(sizeof(double) * 25000000);
  reg_b = malloc(sizeof(double) * 25000000);

  fill(reg_a, reg_b);
  for( int i = 0; i < 25000000; i++){
    reg_a[i] = copysign(reg_a[i], reg_b[i]);
  }
  fill(reg_a, reg_b);
  double start_time = omp_get_wtime();
  for( int i = 0; i < 25000000; i++){
    reg_a[i] = copysign(reg_a[i], reg_b[i]);
  }
  double finish_time = omp_get_wtime();
  printf("copysign took %f\n", (finish_time - start_time));

  fill(reg_a, reg_b);
  start_time = omp_get_wtime();
  for( int i = 0; i < 25000000; i++){
    double z = 0.0;
    if(reg_b[i] >= 0.0) z = 1.0;
    else z = -1.0;
    reg_a[i] = z * reg_a[i];
  }
  finish_time = omp_get_wtime();
  printf("if statements took %f\n", (finish_time - start_time));

  free(reg_a);
  free(reg_b);

}
```

The resulting runtime of this code with 2 different compilers (compiled on Broadwell run on Skylake so running without any chance of AVX512) was:
gcc9 `gcc test_sign.c -fopenmp -O3` :
```
copysign took 0.031722
if statements took 0.033240
```

intel19, `icc test_sign.c -qopenmp -std=c99 -O3`:
```
copysign took 0.035331
if statements took 0.033093
```

There are two takeaways for me here:
1) The regent code takes much longer than the C code. My assumption is that the data layout or something in the mapper makes it slower? The timed regions should otherwise be reasonably optimised by terra as far as I understand? Alternatively the barriers and timing tasks could be slow? But I'd expect that effect to be symmetrical - is there a better way to time sections of regent code?

I could see that the C compilers manage to vectorise the C code for both loops which could give this performance difference perhaps, but that benefit is not seen in Regent.

2) In C code, copysign (regardless of whether the compiler's vectorise the loops or not, and the if statements or not) has performance similar to that of the if statements (if I run repeatedly on broadwell I get performance within 2% either way reliably). In regent, the loops using copysign are consistently  15-20% slower than the loops with the if statement, and consistently 2-3x the performance of the C code. Should the copysign be inlined etc. similar to (my expectation of what happens in) C? Or will there be an additional function call cost that is unavoidable?

**Comment by magnatelee** [comment]
1. I think that big difference you observed is coming from a hidden "mapping" operation that the compiler inserts probably right before the for loop. When a task creates a region, that region is not yet allocated for immediate access but the allocation is deferred until the actual use point. Similarly, results of any fill operations are materialized only when the affected fields are actually read for the first time. So I think what you're seeing is an overhead of mapping and allocating the region for the first time + filling the two fields. The fence you put right before the for loop unfortunately doesn't prevent this factor from being included in the measurement. Can you try to move your loop to a task and time the loop there?

2. I suspect the copysign is not being inlined here. I'll take a look at the disassembly. In any case, I guess the math functions are only LTO-ed, unless Clang and GCC are pulling tricks to handle them specially.

**Comment by magnatelee** [comment]
I was wrong about the copysign; that function seems to be header-only (which I think may not be the case in general for cmath functions) and Terra was able to inline it.

**Comment by LonelyCat124** [comment]
I think the "dummy" loop (thats untimed) I had was partly hiding some of the allocation, but there is definitely an apsect of that. I have tasks as follows now to be timed instead:
```
task novec_copysign( reg : region(ispace(int1d), point), reg2 : region(ispace(int1d), point) )
where writes(reg.a), reads(reg.a, reg2.a) do

  var start_time = c.legion_get_current_time_in_micros()
 __forbid(__vectorize)
 for point in reg do
    reg[point].a = copysign(reg[point].a, reg2[point].a)
 end
  var finish_time = c.legion_get_current_time_in_micros()
  var time_taken = finish_time - start_time
  format.println("non-vectorize copysign: {}", double(time_taken)/1000000.0)
end

task vec_copysign( reg : region(ispace(int1d), point), reg2 : region(ispace(int1d), point) )
where writes(reg.a), reads(reg.a, reg2.a) do
  var start_time = get_time()
  __fence(__execution, __block)
 __demand(__vectorize)
 for point in reg do
    reg[point].a = copysign(reg[point].a, reg2[point].a)
 end
  __fence(__execution, __block)
  var finish_time = get_time()
  var time_taken = finish_time - start_time
  format.println("vectorize copysign: {}", double(time_taken)/1000000.0)
end

task if_copysign(  reg : region(ispace(int1d), point), reg2 : region(ispace(int1d), point) )
where writes(reg.a), reads(reg.a, reg2.a) do
  var start_time = c.legion_get_current_time_in_micros()
  for point in reg do
    var z : double = 0.0
    if reg2[point].a >= 0.0 then
      z = 1.0
    else
      z = -1.0
    end
    reg[point].a = z * reg[point].a
  end
  var finish_time = c.legion_get_current_time_in_micros()
  var time_taken = finish_time - start_time
  format.println("if-statement copysign: {}", double(time_taken)/1000000.0)
end

task main()

 var reg = region( ispace(int1d, 250000000), point)
 var reg2 = region( ispace(int1d, 250000000), point)

 fill(reg.a, 0.75)
 fill(reg2.a, -1.0)

 for point in reg do
    reg[point].a = copysign(reg[point].a, reg2[point].a)
 end

 fill(reg.a, 0.75)
 fill(reg2.a, -1.0)

  __fence(__execution, __block)
  novec_copysign(reg, reg2)
  __fence(__execution, __block)
 fill(reg.a, 0.75)
 fill(reg2.a, -1.0)

  __fence(__execution, __block)
  vec_copysign(reg, reg2)
  __fence(__execution, __block)

 fill(reg.a, 0.75)
 fill(reg2.a, -1.0)


  __fence(__execution, __block)
  if_copysign(reg, reg2)

end
```
In theory now from what you say, the mapping/initialisation/fill operations should be done before the task is launched. 

One strange behaviour I couldn't understand is:
In the `vec_copysign` task, if I directly called `var start_time = c.legion_get_current_time_in_micros()`, then the compiler reported it could no longer vectorise with this error:
```
...ions/regent/legion/language/src/regent/vectorize_loops.t:186: Errors reported during typechecking.
test_sign.rg:37: vectorization failed: loop body has an unsupported function call
    reg[point].a = copysign(reg[point].a, reg2[point].a)
                           ^

stack traceback:
        [C]: in function 'error'
        ...l_versions/regent/legion/language/terra/src/terralib.lua:390: in function 'finishandabortiferrors'
        ...nual_versions/regent/legion/language/src/common/report.t:56: in function 'error'
        ...ions/regent/legion/language/src/regent/vectorize_loops.t:186: in function 'report_error_when_demanded'
        ...ions/regent/legion/language/src/regent/vectorize_loops.t:1147: in function 'expr'
        ...ions/regent/legion/language/src/regent/vectorize_loops.t:737: in function 'stat'
        ...ions/regent/legion/language/src/regent/vectorize_loops.t:677: in function 'block'
        ...ions/regent/legion/language/src/regent/vectorize_loops.t:1231: in function 'fn'
        ..._versions/regent/legion/language/terra/src/terralist.lua:150: in function 'map'
        ...ions/regent/legion/language/src/regent/vectorize_loops.t:1187: in function 'block'
        ...ions/regent/legion/language/src/regent/vectorize_loops.t:1338: in function <...ions/regent/legion/language/src/regent/vectorize_loops.t:1337>
        ...ersions/regent/legion/language/src/regent/passes_hooks.t:41: in function 'optimize'
        ...nual_versions/regent/legion/language/src/regent/passes.t:52: in function <...nual_versions/regent/legion/language/src/regent/passes.t:48>
        test_sign.rg:32: in main chunk
```
I assume there is some small cost associated with the barriers due to this, so I'm not going to focus too heavily on the comparative performance of the vector vs non-vector `copy_sign` results.

Next up performance of regent with this new code:
```
bash-4.2$ legion/language/regent.py test_sign.rg  -ll:csize 4096
non-vectorize copysign: 0.469680
vectorize copysign: 0.539578
if-statement copysign: 0.375238
```

I fixed an error in the C code, and the results of the C code for the same number of iterations is:
gcc9 `gcc test_sign.c -O3 -fopenmp`
```
copysign took 0.364297
if statements took 0.364193
```
icc 19.3 `icc test_sign.c -O3 -qopenmp -std=c99`
```
copysign took 0.353172
if statements took 0.358224
```

icc 19.3 novec `icc test_sign.c -O3 -qopenmp -std=c99 -no-vec`
```
copysign took 0.355829
if statements took 0.367028
```

This looks like regent is doing better for the non-cmath version now, the if statement has comparable performance (within variation) to the C code. However the version using the cmath function still performs much worse. If it would be useful I could compare performance of other cmath functions as well?

Looking at the Intel optimisation report, the intel compiler claims to not vectorise the loops using `copy_sign`, but does vectorise the if statement-based loop (cost of vectorised math library call, after aligning memory with `__mm_malloc`). I assume if there were more heavy arithmetic it may choose to do so but that is not important for this function.

**Comment by elliottslaughter** [comment]
So, just FYI, `__execution_fence` is unnecessary around a leaf task like that. The vectorizer will never generate any tasks, so there is nothing to block on, and meanwhile the fence is not free and probably disables other optimizations such as leaf optimization. I would just `__demand(__vectorize)` and time the loop.

**Comment by LonelyCat124** [comment]
I don't believe the vectorised version is a leaf task as it has to call sub-tasks due to the issue with the vectoriser if I just call the regentlib.c timing function.

I've just tried (and added `__demand(__leaf)` to the other 2 tasks):
```
task vec_copysign( reg : region(ispace(int1d), point), reg2 : region(ispace(int1d), point) )
where writes(reg.a), reads(reg.a, reg2.a) do
  var start_time = get_time()
  __fence(__execution, __block)
 __demand(__vectorize)
 for point in reg do
    reg[point].a = copysign(reg[point].a, reg2[point].a)
 end
  var finish_time = get_time()
  var time_taken = finish_time - start_time
  format.println("vectorize copysign: {}", double(time_taken)/1000000.0)
end
```

and I get this result on SKL:
```
non-vectorize copysign: 0.485700
vectorize copysign: 0.485863
if-statement copysign: 0.330729
```


If you mean to remove the ones around the `get_time()` calls this totally kills the timing, and I have to use the task calls due to the vectoriser complaining if i don't. Its possible that the vectoriser won't vectorise copysign (and is being disabled by the barrier?). If I remove the ones in the main task I get these results (with `-ll:cpu 1 -ll:util 1`, these results are within variance of the above for sure):
```
non-vectorize copysign: 0.486439
vectorize copysign: 0.486633
if-statement copysign: 0.326535
```

For all of these the copysign performance is still worse than the if statements (different to direct C code where the performance is comparable).

**Comment by magnatelee** [comment]
I took a closer look at the disassembly code from the Regent compiler and I confirmed that `copysign` is not being inlined. I also confirmed using clang 6.0 that your C  code has more or less the same performance once I wrap `copysign` in a dummy function and put it in a separate compile unit (which makes the function not inlined).  So I strongly believe that the ~0.15s difference you've observed is due to `copysign` not being inlined, and that both gcc and clang are handling math functions specially. Unless we do something similar in the Regent compiler, which I doubt I could add in the near future, we can't really match performance.

Also note that the Regent compiler cannot vectorize your for loop, since it only vectorizes math functions supported by LLVM intrinsics, such as `abs` or `sqrt`, and AFAIK there is no LLVM intrinsic for `copysign` (that said, I didn't check LLVM releases more recent than 6). So I'm actually surprised that the compiler didn't complain about the pragma; only possibility I can think of is that your tasks weren't leaf tasks, which will turn off the parallelizability checker all together.

**Comment by magnatelee** [comment]
One quick comment: it's certainly possible for Regent to provide its own implementations of math functions that aren't backed by LLVM intrinsics so that they can be vectorized. I just made a [github issue](#864) on this.

**Comment by LonelyCat124** [comment]
Ok - that makes sense, that task couldn't be a leaf task due to the compiler failing on copysign being unvectorisable when it was a leaf task (which I thought was due to the timing calls, but actually was me misunderstanding what Regent was doing).

From my understanding of copysign from the intel compiler's optimisation report, it doesn't have an intrinsic with Intel compiler, the optimsation report has it using a "vectorised math library call" - which is too expensive for this simple loop (but potentially useful for longer loops).

**Comment by magnatelee** [comment]
> Ok - that makes sense, that task couldn't be a leaf task due to the compiler failing on copysign being unvectorisable when it was a leaf task (which I thought was due to the timing calls, but actually was me misunderstanding what Regent was doing).

You can make sure your task is a leaf by putting `__demand(__leaf)`.

> From my understanding of copysign from the intel compiler's optimisation report, it doesn't have an intrinsic with Intel compiler, the optimsation report has it using a "vectorised math library call" - which is too expensive for this simple loop (but potentially useful for longer loops).

Yeah it doesn't need to be an intrinsic to do the job. Compilers just need to syntactically recognize math functions and do something special on them.

> Edit: while googling some related stuff I came across this in the llvm 10 docs: https://llvm.org/docs/LangRef.html#llvm-copysign-intrinsic - is this the intrinsic Regent would need in the future? (same link for 5.x https://releases.llvm.org/5.0.2/docs/LangRef.html#llvm-copysign-intrinsic )

Although those intrinsics certainly help to get math functions inlined for non-vectorized case, those are not vector intrinsics, so Regent can't really use them in the vectorizer. There are lower level intrinsics, such as `llvm.x86.avx.sin.pd.256` that are directly mapped to vector instructions.
